url11fb0973cf_raichuman9_Swimmer526_tgz-pJ8-fr_inria_optimization_cmaes_examples_CMAExample1J8/src/ubc/swim/optimization/SwimmerOptimization.java

```diff
@@ -0,0 +1,168 @@
+
+		// Write output file initial header
+		cma.writeToDefaultFilesHeaders(0); // 0 == overwrites old files
+
+		// Run optimization iterations until a stop condition is met
+		while(cma.stopConditions.getNumber() == 0) {
+			//Get a list of sampled control strategies
+			double[][] controlPop = cma.samplePopulation(); 
+			
+			//Check fitness of each strategy
+			for (int i = 0; i < controlPop.length; ++i) {    
+				//If strategy is disallowed, resample (watch for infinite loops here...)
+				while (!fitFun.isFeasible(controlPop[i]))  
+					controlPop[i] = cma.resampleSingle(i);    
+                                                     
+                // compute fitness/objective value (to be minimized)
+				fitness[i] = fitFun.valueOf(controlPop[i]); 
+			}
+			
+			// Pass fitness array to update search distribution
+			cma.updateDistribution(fitness);         
+
+			// Update output to console/files
+			cma.writeToDefaultFiles();
+			//Print table headers and stats every so often
+			if (cma.getCountIter() % (15*iterationsPerOutput) == 1)
+				cma.printlnAnnotation(); 
+			if (cma.getCountIter() % iterationsPerOutput == 1)
+				cma.println(); 
+		}
+		
+		// Set best-ever solution to mean control strategy (which is best guess for optimal strategy)
+		cma.setFitnessOfMeanX(fitFun.valueOf(cma.getMeanX())); 
+
+		// Write final output
+		cma.writeToDefaultFiles(1);
+		cma.println();
+		cma.println("Terminated due to");
+		for (String s : cma.stopConditions.getMessages())
+			cma.println("  " + s);
+		cma.println("best function value " + cma.getBestFunctionValue() 
+				+ " at evaluation " + cma.getBestEvaluationNumber());
+		
+		return cma.getBestX();
+	}
+	
+	/**
+	 * Writes given control values to a comma-separated value file with given path and name
+	 * (.csv extension will be automatically appended)
+	 * @param x
+	 * @param filePath location folder of file (no trailing "/")
+	 * @param fileName name of file (no .csv extension)
+	 */
+	public static void writeToCSV(double[] x, String filePath, String fileName) {
+		try
+		{
+		    FileWriter writer = new FileWriter(filePath + "/" + fileName + ".csv");
+		    
+		    for (double val : x) {
+		    	writer.append(Double.toString(val));
+		    	writer.append(",\n");
+		    }
+	 
+		    writer.flush();
+		    writer.close();
+		}
+		catch(IOException e)
+		{
+		     e.printStackTrace();
+		} 
+	}
+	
+	/**
+	 * Reads control values from a comma-separated value file with given path
+	 * and name and returns them in a double array (.csv extension will be
+	 * automatically appended)
+	 * 
+	 * @param filePath
+	 *            location folder of file (no trailing "/")
+	 * @param fileName
+	 *            name of file (no .csv extension)
+	 */
+	public static double[] readFromCSV(String filePath, String fileName) {
+		List<Double> loadedVals = new ArrayList<Double>();
+		
+		try
+		{
+			BufferedReader reader  = new BufferedReader(new FileReader(filePath + "/" + fileName + ".csv"));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *reader*. The resource is referenced at lines 143, 147. The resource is closed at line 153. There are other execution paths that do not contain closure statements, for example, when BufferedReader.readLine throws an exception. Ensure *reader* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).